pkg/operator/ceph/object/controller.go

```diff
@@ -181,12 +218,50 @@ func (c *ObjectStoreController) onDelete(obj interface{}) {
 	c.acquireOrchestrationLock()
 	defer c.releaseOrchestrationLock()
 
+	err = c.handleDelete(objectstore)
+	if err != nil {
+		logger.Errorf("failed to delete objectstore. %v", err)
+	}
+
 	cfg := clusterConfig{context: c.context, store: *objectstore}
 	if err = cfg.deleteStore(); err != nil {
 		logger.Errorf("failed to delete object store %q. %v", objectstore.Name, err)
 	}
 }
 
+func (c *ObjectStoreController) handleDelete(objectstore *cephv1.CephObjectStore) error {
+	for retryCount := 0; retryCount < objectstoreDeleteMaxRetries; retryCount++ {
+
+		selector := fmt.Sprintf("metadata.namespace=%s", objectstore.Namespace)
+		objectBuckets, err := c.bktclient.ObjectbucketV1alpha1().ObjectBuckets().List(metav1.ListOptions{FieldSelector: selector})
+		if err != nil {
+			return errors.Wrap(err, "failed to get buckets for objectstore %q in namespace %q")
+		}
+		if len(objectBuckets.Items) == 0 {
+			logger.Infof("no objectbuckets for objectstore %q to clean up.", objectstore.Namespace)
+			break
+		}
+
+		retryCount++
+		if retryCount == objectstoreDeleteMaxRetries {
+			bucketNames := make([]string, 0)
+			for _, bucket := range objectBuckets.Items {
+				bucketNames = append(bucketNames, bucket.Name)
+			}
+			logger.Warningf(
+				"exceeded retry count while waiting for buckets for objectstore %q to be cleaned up. buckets: %+v",
+				objectstore.Namespace,
+				bucketNames)
+			break
+		}
+
+		logger.Infof("waiting for buckets in objectstore %q to be cleaned up. Retrying in %q.",
+			objectstore.Namespace, objectstoreDeleteRetryInterval.String())
+		<-time.After(objectstoreDeleteRetryInterval)
+	}
+	return nil
```

Comment:
Should we return an err if we timeout while waiting for the buckets to be purged? Otherwise we will go ahead and remove the finalizer even while the buckets still exist.

Reply:
I don't disagree with the sentiment that we should return an error. However, this patch is intentionally mirroring the logic in the CSI/FlexVolume cleanup finalizer of the ClusterController.

I think that once Seb's controller patches are in, we should definitely change all of the above to return an error. In the meantime, I can see both arguments, and historically we've rolled over similar errors. As per our conversation off the PR, we'll leave this logic as is for now.